File: gn3/computations/biweight.R

library(WGCNA)

FetchArgs <- function(){
	myArgs <- commandArgs(trailingOnly = TRUE)
	trait_vals <- as.numeric(unlist(strsplit(myArgs[1], split=" ")))
	target_vals <- as.numeric(unlist(strsplit(myArgs[2], split=" ")))

	return(list(trait_vals= c(trait_vals),target_vals = c(target_vals)))

}
BiweightMidCorrelation <- function(trait_val,target_val){

	results <- bicorAndPvalue(c(trait_val),c(target_val))
    return ((c(c(results$bicor)[1],c(results$p)[1])))
}


results <- (BiweightMidCorrelation(FetchArgs()[1],FetchArgs()[2]))

cat(results)

File: gn3/computations/biweight.py
"""module contains script to call biweight midcorrelation in R"""
import subprocess

from typing import List
from typing import Tuple

from gn3.settings import BIWEIGHT_RSCRIPT


def calculate_biweight_corr(trait_vals: List,
                            target_vals: List,
                            path_to_script: str = BIWEIGHT_RSCRIPT,
                            command: str = "Rscript"
                            ) -> Tuple[float, float]:
    """biweight function"""

    args_1 = ' '.join(str(trait_val) for trait_val in trait_vals)
    args_2 = ' '.join(str(target_val) for target_val in target_vals)
    cmd = [command, path_to_script] + [args_1] + [args_2]

    results = subprocess.check_output(cmd, universal_newlines=True)

    (corr_coeff, p_val) = tuple([float(y) for y in results.split()])

    return (corr_coeff, p_val)
